### PR TITLE
Requirements upgrade November 2024

### PR DESCRIPTION
**Pull request by amercader**
Upgrading requirements to the latest versions. No major issues:

* Upgraded Werkzeug to 3.1: some minor changes in how some edge-case params are rendered
* markupsafe / markdown / bleach: not sure which upgrade exactly caused it but there were some minor changes in the html output of some helpers
* Sphinx / sphinx-rtd-theme: minor changes in docs configuration

### CODE REVIEW

**Comment by amercader** [comment]
This and #8505 will clear some 
vulnerabilities listed on Snyk:

Screenshot 2024-11-05 at 13-23-55 Projects Snyk

File: ckan/lib/helpers.py [hunk 1]
```diff
@@ -451,7 +453,7 @@ def _url_for_flask(*args: Any, **kw: Any) -> str:
                 for key, val in kw.items():
                     if isinstance(val, (list, tuple)):
                         for value in val:
-                            if value is None:
+                            if value is None or (isinstance(val, dict) and not val):
```
Comment (wardi) [inline ckan/lib/helpers.py:456]: Is there a reason not to
```suggestion
                            if value is None or val == {}:
```

**Review by wardi — COMMENTED** [review]
Just that minor style comment, otherwise LGTM

File: ckan/lib/helpers.py [hunk 2]
```diff
@@ -460,14 +462,15 @@ def _url_for_flask(*args: Any, **kw: Any) -> str:
                                 )
                             )
                     else:
-                        if val is None:
+                        if val is None or (isinstance(val, dict) and not val):
```
Comment (wardi) [inline ckan/lib/helpers.py:465]:
```suggestion
                        if val is None or val == {}:
```

**Comment by amercader** [comment]
I'll put this on hold as the pytest upgrades break extensions tests due to a problem with the `pytest-ckan` package

File: ckan/lib/helpers.py (hunk 1 above)
Comment (wardi) [reply ckan/lib/helpers.py:456]: @smotornyuk Do you know a workaround for this situation?

Comment (amercader) [reply ckan/lib/helpers.py]: My bad, I confused the `value` and `val` variables, now it makes sense: [6030628](https://github.com/ckan/ckan/pull/8504/commits/6030628cf2a02d59d897fccddc972852e80cd739)

**Comment by amercader** [comment]
@wardi I've created separate issues for the pytest thing, this is ready for final review

**Pull request merged**